File: libgomp/testsuite/libgomp.c-c++-common/declare-mapper-13.c
/* { dg-do run } */

#include <assert.h>

struct T {
  int a;
  int b;
  int c;
};

void foo (void)
{
  struct T x;
  x.a = x.b = x.c = 0;

#pragma omp target
  {
    x.a++;
    x.c++;
  }

  assert (x.a == 1);
  assert (x.b == 0);
  assert (x.c == 1);
}

// An identity mapper.  This should do the same thing as the default!
#pragma omp declare mapper (struct T v) map(v)

void bar (void)
{
  struct T x;
  x.a = x.b = x.c = 0;

#pragma omp target
  {
    x.b++;
  }

#pragma omp target map(x)
  {
    x.a++;
  }

  assert (x.a == 1);
  assert (x.b == 1);
  assert (x.c == 0);
}

int main (int argc, char *argv[])
{
  foo ();
  bar ();
  return 0;
}

File: libgomp/testsuite/libgomp.fortran/declare-mapper-13.f90
! { dg-do run }

module mymod
type S
integer :: a
integer :: b
integer :: c
end type S

!$omp declare mapper (S :: x) map(x%c)
end module mymod

program myprog
use mymod
type T
integer :: a
integer :: b
integer :: c
end type T

type(S) :: mys
type(T) :: myt

!$omp declare mapper (T :: x) map(x%b)

myt%a = 0
myt%b = 0
myt%c = 0
mys%a = 0
mys%b = 0
mys%c = 0

!$omp target
myt%b = myt%b + 1
!$omp end target

!$omp target
mys%c = mys%c + 1
!$omp end target

!$omp target
myt%b = myt%b + 2
mys%c = mys%c + 3
!$omp end target

if (myt%b.ne.3) stop 1
if (mys%c.ne.4) stop 2

end program myprog
